File: mobile/modules/native-upload/android/src/main/java/expo/modules/nativeupload/NativeUploadModule.kt
package expo.modules.nativeupload

import android.content.Intent
import androidx.core.content.ContextCompat
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

/**
 * 진짜 백그라운드 업로드 — 포그라운드 서비스에 위임(앱 닫혀도 dataSync 서비스가 완료).
 * 압축은 Media3 Transformer(오디오 passthrough), 업로드는 청크+resume. 완료 durability는
 * 서버 record-first-patch(target_type/target_id) + 알림. 이벤트는 앱이 살아있을 때 best-effort.
 */
class NativeUploadModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NativeUpload")

    Events("uploadProgress", "uploadComplete")

    OnCreate {
      sink = { name, body -> this@NativeUploadModule.sendEvent(name, body) }
    }
    OnDestroy {
      sink = null
    }

    Function("isAvailable") { true }

    AsyncFunction("requestPermissions") { promise: Promise ->
      // POST_NOTIFICATIONS는 매니페스트 선언 + 런타임 요청은 JS(expo-notifications)나 최초 실행에 위임.
      // 서비스 자체는 알림 없이도 동작하므로 여기선 선언 상태를 반환.
      promise.resolve(mapOf("notifications" to true))
    }

    AsyncFunction("enqueueUpload") { options: Map<String, Any?>, promise: Promise ->
      val ctx = appContext.reactContext
      if (ctx == null) { promise.reject("E_CTX", "no android context", null); return@AsyncFunction }
      val id = UUID.randomUUID().toString()
      val intent = Intent(ctx, UploadForegroundService::class.java).apply {
        putExtra("id", id)
        putExtra("fileUri", options["fileUri"] as? String)
        putExtra("apiUrl", options["apiUrl"] as? String)
        putExtra("token", options["token"] as? String)
        putExtra("subfolder", (options["subfolder"] as? String) ?: "portfolios")
        putExtra("targetType", options["targetType"] as? String)
        putExtra("targetId", options["targetId"] as? String)
        putExtra("displayName", (options["displayName"] as? String) ?: "영상")
        putExtra("compress", (options["compress"] as? Boolean) ?: true)
      }
      ContextCompat.startForegroundService(ctx, intent)
      promise.resolve(mapOf("enqueued" to true, "id" to id))
    }
  }

  companion object {
    @Volatile
    private var sink: ((String, Map<String, Any?>) -> Unit)? = null

    /** 서비스에서 JS로 이벤트 전달(앱 살아있을 때만). */
    fun emit(name: String, body: Map<String, Any?>) {
      sink?.invoke(name, body)
    }
  }
}

File: mobile/modules/native-upload/ios/NativeUploadModule.swift
import ExpoModulesCore
import Foundation
import UserNotifications

public class NativeUploadModule: Module {
    public func definition() -> ModuleDefinition {
        Name("NativeUpload")

        Events("uploadProgress", "uploadComplete")

        OnCreate {
            BackgroundUploader.shared.activate()
            BackgroundUploader.shared.onProgress = { [weak self] id, p in
                self?.sendEvent("uploadProgress", ["id": id, "phase": "uploading", "progress": p])
            }
            BackgroundUploader.shared.onComplete = { [weak self] id, ok, url, err in
                var payload: [String: Any] = ["id": id, "ok": ok]
                if let u = url { payload["url"] = u }
                if let e = err { payload["error"] = e }
                self?.sendEvent("uploadComplete", payload)
            }
        }

        Function("isAvailable") { () -> Bool in true }

        AsyncFunction("requestPermissions") { (promise: Promise) in
            UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, _ in
                promise.resolve(["notifications": granted])
            }
        }

        AsyncFunction("enqueueUpload") { [weak self] (options: [String: Any], promise: Promise) in
            guard let fileUri = options["fileUri"] as? String,
                  let apiUrl = options["apiUrl"] as? String,
                  let token = options["token"] as? String else {
                promise.reject("E_ARGS", "fileUri/apiUrl/token 필수"); return
            }
            let subfolder = options["subfolder"] as? String ?? "portfolios"
            let targetType = options["targetType"] as? String
            let targetId = options["targetId"] as? String
            let displayName = options["displayName"] as? String ?? "영상"
            let compress = options["compress"] as? Bool ?? true

            let id = UUID().uuidString
            let origPath = fileUri.hasPrefix("file://") ? (URL(string: fileUri)?.path ?? fileUri) : fileUri
            let origExt = (origPath as NSString).pathExtension.lowercased()
            let videoExts = ["mp4", "mov", "m4v", "3gp", "avi", "mkv", "webm", "hevc"]
            let isVideo = videoExts.contains(origExt)

            func mimeAndName(for uploadPath: String) -> (String, String) {
                let base = ((origPath as NSString).lastPathComponent as NSString).deletingPathExtension
                if isVideo { return ("video/mp4", base.isEmpty ? "video.mp4" : base + ".mp4") }
                let imageExts = ["jpg", "jpeg", "png", "heic", "heif", "webp", "gif"]
                let ext = (uploadPath as NSString).pathExtension.lowercased()
                if imageExts.contains(ext) {
                    return ("image/\(ext == "jpg" ? "jpeg" : ext)", (origPath as NSString).lastPathComponent)
                }
                return ("application/octet-stream", (origPath as NSString).lastPathComponent)
            }

            func finishEnqueue(_ uploadPath: String) {
                let (mime, name) = mimeAndName(for: uploadPath)
                let ok = BackgroundUploader.shared.enqueue(
                    id: id, filePath: uploadPath, apiUrl: apiUrl, token: token,
                    subfolder: subfolder, targetType: targetType, targetId: targetId,
                    mimeType: mime, uploadFileName: name, displayName: displayName)
                if ok { promise.resolve(["enqueued": true, "id": id]) }
                else { promise.reject("E_ENQUEUE", "업로드 큐 등록 실패") }
            }

            if compress && isVideo {
                self?.sendEvent("uploadProgress", ["id": id, "phase": "compressing", "progress": 0])
                VideoCompressor.compress(inputPath: origPath, onProgress: { p in
                    self?.sendEvent("uploadProgress", ["id": id, "phase": "compressing", "progress": p])
                }, completion: { outPath in
                    finishEnqueue(outPath ?? origPath)  // nil → 오디오소실/실패 → 원본
                })
            } else {
                finishEnqueue(origPath)
            }
        }
    }
}
